Fix linalg.norm sample inputs function and related failures

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84631
* #84455
* #84567
* #84454
* #84554
* #84696
* #84453
* __->__ #84452

Due to an indentation error, the return statement happens after just 1
loop of `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

peterbell10 committed:
Fix linalg.norm sample inputs function and related failures

Due to an indentation error, the return statement happens after just 1
loop of the `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of the `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of the `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of the `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of the `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

Lezcano commented on torch/testing/_internal/opinfo/definitions/linalg.py:
```diff
@@ -555,6 +558,28 @@ def sample_inputs_linalg_norm(
+                        -np.inf: (0,),
+                    }.get(ord, ())
+
+                    if any(test_size[d] == 0 for d in dims_to_check):
+                        # IndexError: amax(): Expected reduction dim {dim} to
+                        # have non-zero size.
```
This doesn't happen in the `"fro"` and `"nuc"` cases.

Lezcano commented on torch/testing/_internal/opinfo/definitions/linalg.py:
```diff
@@ -555,6 +558,28 @@ def sample_inputs_linalg_norm(
+                        # does not have an identity
+                        continue
+                elif is_matrix_norm:
+                    dims_to_check = {
+                        None: (0,),
+                        np.inf: (0,),
+                        2: (0, 1),
+                        1: (1,),
+                        -1: (1,),
+                        -2: (0, 1),
+                        -np.inf: (0,),
+                    }.get(ord, ())
```
Are `nuc` and `fro` not considered here? And if so, why?

peterbell10 replied:
The `fro` and `nuc` cases work for all sizes, so we go to the default `()` i.e. not checking any dims. I could spell it out explicitly if you prefer.

Lezcano replied:
Ah, right. For some reason I thought we were skipping that case.

lezcano reviewed (commented):
Uf, good catch. 
That function could do with some refactoring by using `make_arg` and turning it into a generator for clarity.

I left a couple comments, but otherwise it looks good. I can put up a follow-up PR fixing all these xfails.

Lezcano commented:
fwiw, for a good high-level summary of what this function does, see the reference implementation at https://github.com/pytorch/pytorch/blob/master/torch/_refs/linalg/__init__.py

lezcano reviewed (approved):
Cool, thanks!

Lezcano commented:
Could we have an OK on this one @zou3519? My linalg powers are not enough to get this one merged as it touches functorch.

peterbell10 committed:
Update on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 commented on functorch/test/test_vmap.py:
```diff
@@ -3304,6 +3304,8 @@ def test_vmap_exhaustive(self, device, dtype, op):
         xfail('linalg.matrix_rank'),
         xfail('linalg.matrix_rank', 'hermitian'),
+        xfail('linalg.norm'),
+        xfail('linalg.norm', 'subgradients_at_zero'),
```
@zou3519 PTAL, I've applied your fix.

zou3519 replied:
LGTM, thanks for applying the fix

peterbell10 commented:
@pytorchbot merge -l

peterbell10 committed:
Update on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "Fix linalg.norm sample inputs function and related failures"


Due to an indentation error, the return statement happens after just 1
loop of `for test_size in test_sizes` so only one shape was ever
tested.

This also revealed several cases where the provided shapes don't work
so I've disabled the generation of those sample inputs.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -f "macos has unrelated failure during homebrew setup"
